File: typescript/tests/booking/application/integration/payment/adapter/paymentRequest.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { bookingPlaced } from '@/booking/domain/events/bookingPlaced.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import { paymentRequestFromBookingPlaced } from '@/booking/application/integration/payment/adapter/paymentRequest.js';

describe('PaymentRequest (booking -> payment)', () => {
  it('adapts BookingPlaced to PaymentRequestIntegrationCommand', () => {
    const bookingId = generateId((value) => new BookingId(value));
    const amount = new Money(75);
    const description = new Description('suite');
    const giftCardReference = crypto.randomUUID();
    const event = bookingPlaced(bookingId, amount, description, giftCardReference);

    const command = paymentRequestFromBookingPlaced(event);

    expect(command.clientReference).toBe(bookingId.value.value);
    expect(command.amount).toBe(amount);
  });

  it('throws on null event', () => {
    expect(() => paymentRequestFromBookingPlaced(null as unknown as ReturnType<typeof bookingPlaced>)).toThrow();
  });
});

File: java/src/test/java/it/giannibombelli/wsc2026/booking/application/integration/payment/adapter/PaymentRequestTest.java
package it.giannibombelli.wsc2026.booking.application.integration.payment.adapter;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.events.BookingPlaced;
import it.giannibombelli.wsc2026.payment.integration.PaymentRequestIntegrationCommand;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PaymentRequestTest {

    @Test
    void fromBookingPlaced_returnsIntegrationCommand() {
        BookingId bookingId = EntityId.generate(BookingId::new);
        Money amount = new Money(new BigDecimal("75.00"));
        Description description = new Description("suite");
        UUID giftCardReference = UUID.randomUUID();
        BookingPlaced event = new BookingPlaced(bookingId, amount, description, giftCardReference);

        PaymentRequestIntegrationCommand command = PaymentRequest.fromBookingPlaced(event);

        assertThat(command.clientReference()).isEqualTo(bookingId.value().toString());
        assertThat(command.amount()).isEqualTo(amount);
    }

    @Test
    void fromBookingPlaced_withNullEvent_throwsException() {
        assertThatThrownBy(() -> PaymentRequest.fromBookingPlaced(null))
            .isInstanceOf(IllegalArgumentException.class);
    }
}
